src/tokens/selectors.ts

```diff
@@ -42,21 +43,97 @@ const DEFAULT_MEMOIZE_MAX_SIZE = 10
 
 function isNetworkIdList(networkIds: any): networkIds is NetworkId[] {
   return (
-    networkIds.constructor === Array &&
+    Array.isArray(networkIds) &&
     networkIds.every((networkId) => Object.values(NetworkId).includes(networkId))
   )
 }
 export const tokenFetchLoadingSelector = (state: RootState) => state.tokens.loading
 export const tokenFetchErrorSelector = (state: RootState) => state.tokens.error
 
+// Note: not importing from 'src/positions/selectors' to avoid circular dependency
+// TODO: address circular dependency
+const positionsSelector = (state: RootState) => state.positions.positions
+const positionsFetchedAtSelector = (state: RootState) => state.positions.positionsFetchedAt
+
+const positionTokensSelector = createSelector([positionsSelector], (positions) => {
+  const positionTokens: Record<string, Token> = {}
+
+  function visitToken(token: Token, isTopLevelAppToken: boolean) {
+    const tokenId = token.tokenId
+    if (!positionTokens[tokenId] || isTopLevelAppToken) {
+      positionTokens[tokenId] = {
+        ...token,
+        // Only keep the balance if it's a top level app-token
+        // Otherwise it doesn't represent a real user balance
+        balance: isTopLevelAppToken ? token.balance : '0',
+      }
+    }
+    if (token.type === 'app-token') {
+      for (const childToken of token.tokens) {
+        visitToken(childToken, false)
+      }
+    }
+  }
+
+  for (const position of positions) {
+    if (position.type === 'app-token') {
+      visitToken(position, true)
+    } else {
+      for (const token of position.tokens) {
+        visitToken(token, false)
+      }
+    }
+  }
+  return positionTokens
+})
+
 export const tokensByIdSelector = createSelector(
   [
     (state: RootState) => state.tokens.tokenBalances,
+    positionTokensSelector,
+    positionsFetchedAtSelector,
     (_state: RootState, networkIds: NetworkId[]) => networkIds,
   ],
-  (storedBalances, networkIds) => {
+  (storedBalances, positionTokens, positionsFetchedAt, networkIds) => {
+    const allStoredBalances = { ...storedBalances }
+
+    // Enrich with position tokens
+    // This allows us to have priceUsd and balance for tokens
+    // decomposed via positions
+    for (const positionToken of Object.values(positionTokens)) {
```

Comment:
Could we destructure here and adjust the allStoredBalances[tokenId] values to use the new consts? This avoids the `tokenId` and `priceUsd` consts.

```TypeScript
for (const {
  priceUsd,
  tokenId,
  address,
  networkId,
  decimals,
  symbol,
  balance,
} of Object.values(positionTokens)) {
```

Reply:
Yes we could, I didn't do it to make it a bit clearer as to what is coming from `positionToken` and what is `existingToken`.